OpenRA.Game/Network/Order.cs

```diff
@@ -310,6 +316,10 @@ public byte[] Serialize()
 
 			switch (Type)
 			{
+				case OrderType.Disconnect:
+					w.Write(ExtraData);
```

Comment:
What is this ExtraData used for?

Reply:
Used for sending the ClientId.

Reply:
S->C is always Type+Length+ClientId+Frame so you already have the ClientId there. You don't want it in the C->S direction anyway, especially if the goal of this PR is to avoid spoofing.

Reply:
Not in this case, this is sent from the server, before the server did spoof this as sent from the Client. See https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/Server/Server.cs#L1042

Reply:
Then I don't really understand the goal of this PR. I don't really see the current logic as spoofing, because the server simply translate the EOF (which comes from the client, directly or indirectly) into something that can be sent to other clients. It is also a dedicated frame type (and not an order) and those are meant for such special cases in my opinion.

Note that there is (was?) also a Disconnect order which was sent as a normal order over the connection. I think it would be much better to remove that one instead, so all disconnect situations work exactly the same way and avoid the *order* spoofing situation.

Reply:
Calling it spoofing was maybe using the wrong word. more like stop faking that the disconnect order is sent from the client, and the goal of this PR is to clean up the network code.

Reply:
How do the other clients know when (which frame) the disconnect happened? Server orders are frame=0 by definition and you don't have a dedicated field for it.

Reply:
I have always seen the protocol like this, and then the current design makes sense (except for the disconnect *order* which was introduced later):
- A client can only send stuff from itself, so the ClientId field is omitted for C->S. Otherwise it works the same as S->C.
- The server basically bundles many logical streams (one per client plus one for the server) into a single connection:
  - The server is just another client, but with ClientId=0.
  - Frames with Frame=0 are to be processed immediately.
  - All other frames form an ordered stream based on Frame!=0, which must not have any gaps or duplicates.
  - A new stream is implicitely opened if the ClientId hasn't been seen before.
  - A stream is closed explicitely with a Disconnect frame (plus a Disconnect order which is weird).
  - In case of replays the Disconnect frame and order are ignored and it is derived from the last encountered order (weird).
  - SyncHash is optional data and ordered independently. It would probably make more sense to set Frame=0 and include the real frame number inside the payload? On the other hand it is also a stream, just independent from the order stream.

Reply:
>  How do the other clients know when (which frame) the disconnect happened? Server orders are frame=0 by definition and you don't have a dedicated field for it.
>

It was never needed, and the code was just weird and was removed in a previous PR.

> 
> I have always seen the protocol like this, and then the current design makes sense (except for the disconnect _order_ which was introduced later):
> 
>     * A client can only send stuff from itself, so the ClientId field is omitted for C->S. Otherwise it works the same as S->C.
> 
>     * The server basically bundles many logical streams (one per client plus one for the server) into a single connection:
>       
>       * The server is just another client, but with ClientId=0.

Correct.

>       * Frames with Frame=0 are to be processed immediately.

Correct.

>       * All other frames form an ordered stream based on Frame!=0, which must not have any gaps or duplicates.
>       * A new stream is implicitely opened if the ClientId hasn't been seen before.

Not anymore, now these are created when the games is started.

>       * A stream is closed explicitely with a Disconnect frame (plus a Disconnect order which is weird).

In the future i want to remove one of these orders.

>       * In case of replays the Disconnect frame and order are ignored and it is derived from the last encountered order (weird).

Something for a future cleanup.

>       * SyncHash is optional data and ordered independently. It would probably make more sense to set Frame=0 and include the real frame number inside the payload? On the other hand it is also a stream, just independent from the order stream.